content/client-server-api/modules/event_replacements.md

```diff
@@ -0,0 +1,319 @@
+
+ * Neither the replacement nor original events may have a `state_key` property
+   (i.e. it is not possible to edit a state event).
+
+ * The original event must not, itself, have a `rel_type` of `m.replace`
+   (i.e. you cannot edit an edit — though you can send multiple edits for a
+   single original event).
+
+ * The replacement event (once decrypted, if appropriate) must have an
+   `m.new_content` property.
+
+If any of these criteria are not satisfied, implementations should ignore the
+replacement event (the content of the original should not be replaced, and the
+edit should not be included in the server-side aggregation).
+
+Note that the [`msgtype`](#mroommessage-msgtypes) property of replacement
+`m.room.message` events need *not* be the same as in the original event. For
+example, it is legitimate to replace an `m.text` event with an `m.emote`.
+
+#### Editing encrypted events
+
+If the original event was [encrypted](#end-to-end-encryption), the replacement
+should be too. In that case, `m.new_content` is placed in the content of the
+encrypted payload. As with all event relationships, the `m.relates_to` property
+must be sent in the unencrypted (cleartext) part of the event.
+
+For example, a replacement for an encrypted event might look like this:
+
+For example, an encrypted replacement event might look like this:
+
+```json
+{
+    "type": "m.room.encrypted",
+    "content": {
+        "m.relates_to": {
+            "rel_type": "m.replace",
+            "event_id": "$some_event_id"
+        },
+        "algorithm": "m.megolm.v1.aes-sha2",
+        "sender_key": "<sender_curve25519_key>",
+        "device_id": "<sender_device_id>",
+        "session_id": "<outbound_group_session_id>",
+        "ciphertext": "<encrypted_payload_base_64>"
+    }
+    // ...
+}
+```
+
+... and, once decrypted, the payload might look like this:
+
+```json
+{
+    "type": "m.room.<event_type>",
+    "room_id": "!some_room_id",
+    "content": {
+        "body": "* Hello! My name is bar",
+        "msgtype": "m.text",
+        "m.new_content": {
+            "body": "Hello! My name is bar",
+            "msgtype": "m.text"
+        }
+    }
+}
+```
+
+Note that:
+
+ * There is no `m.relates_to` property in the encrypted payload. (Any such property would be ignored.)
+ * There is no `m.new_content` property in the cleartext content of the `m.room.encrypted` event. (Again, any such property would be ignored.)
+
+The payload of an encrypted replacement event must be encrypted as normal, including ratcheting any [Megolm](#mmegolmv1aes-sha2) session as normal. The original Megolm ratchet entry should **not** be re-used.
+
+
+#### Applying `m.new_content`
+
+When applying a replacement, the `content` of the original event is replaced
+entirely by the `m.new_content` from the replacement event, with the exception
+of `m.relates_to`, which is left *unchanged*. (Any `m.relates_to` property
+within `m.new_content` should be ignored.)
```

Comment:
This says "replace" a lot:

```suggestion
When applying a replacement, the `content` of the original event is treated as
overwritten entirely by `m.new_content`, with the exception of `m.relates_to`,
which is left *unchanged* because any `m.relates_to` property within `m.new_content`
is ignored. Note that the original event's `content` is not actually overwritten
or replaced by the new content: this is to provide history of the changes.
```

also added a bit to clarify that the original event doesn't change because that'd cause problems for auth and such: we're just dealing with it at the client level like that, because we can.

Reply:
One of the ideas here is that client impls *should* be allowed to overwrite the original event's `content`. You're right that servers need to preserve the original though.

```suggestion
When applying a replacement, the `content` of the original event is treated as
being overwritten entirely by `m.new_content`, with the exception of `m.relates_to`,
which is left *unchanged*. Any `m.relates_to` property within `m.new_content`
is ignored. (Note that server implementations must not *actually* overwrite
the original event's `content`: instead the server presents it as being overwritten
when it is served over the client-server API. See [Server-side replacement of content](#server-side-replacement-of-content)
below.
```